clang/lib/Sema/SemaStmt.cpp

```diff
@@ -204,23 +205,29 @@ static bool DiagnoseUnusedComparison(Sema &S, const Expr *E) {
   return true;
 }
 
-static bool DiagnoseNoDiscard(Sema &S, const WarnUnusedResultAttr *A,
-                              SourceLocation Loc, SourceRange R1,
-                              SourceRange R2, bool IsCtor) {
+static bool DiagnoseNoDiscard(Sema &S, const NamedDecl *OffendingDecl,
+                              const WarnUnusedResultAttr *A, SourceLocation Loc,
+                              SourceRange R1, SourceRange R2, bool IsCtor) {
   if (!A)
     return false;
   StringRef Msg = A->getMessage();
 
+  bool result;
   if (Msg.empty()) {
     if (IsCtor)
-      return S.Diag(Loc, diag::warn_unused_constructor) << A << R1 << R2;
-    return S.Diag(Loc, diag::warn_unused_result) << A << R1 << R2;
-  }
+      result = S.Diag(Loc, diag::warn_unused_constructor) << A << R1 << R2;
+    else
+      result = S.Diag(Loc, diag::warn_unused_result) << A << R1 << R2;
+  } else if (IsCtor)
+    result = S.Diag(Loc, diag::warn_unused_constructor_msg)
+             << A << Msg << R1 << R2;
+  else
+    result = S.Diag(Loc, diag::warn_unused_result_msg) << A << Msg << R1 << R2;
 
-  if (IsCtor)
-    return S.Diag(Loc, diag::warn_unused_constructor_msg) << A << Msg << R1
-                                                          << R2;
-  return S.Diag(Loc, diag::warn_unused_result_msg) << A << Msg << R1 << R2;
+  if (OffendingDecl)
```

Comment:
Why is this allowed to be null?  What cases do we have where this doesn't have a declaration?

Reply:
Well, the initial motivation for that if is the following call site:
```cpp
    const NamedDecl *OffendingDecl;
    const Attr *A;
    std::tie(OffendingDecl, A) = CE->getUnusedResultAttr(Context);
    if (DiagnoseNoDiscard(*this, OffendingDecl,
                          cast_or_null<WarnUnusedResultAttr>(A), Loc, R1, R2,
                          /*isCtor=*/false))
      return;
```
In `CallExpr::getUnusedResultAttr` we tried to see if the callee is marked nodiscard and if so, returns that instead. However `getCalleeDecl()` returns a `const Decl *`, but the diagnostic requires a `const NamedDecl *`, so I dyn_cast it down. If the actual dynamic type is always NamedDecl or its subclass, maybe here the return value cannot ever be `nullptr` to begin with? I was not sure about this so I added the check.

Reply:
I can't imagine anything that would not be a `NamedDecl` that one could put the `nodiscard` on.  Also though, WHY dose it have to be a `NamedDecl`?  We only need the location for the diagnostic anyway, so having `getUnusedResultAttr` return just a `Decl` is probably sufficient.

Reply:
Ah, woops!  I missed that you were doing a print of the name in the note.  I'd suggest doing a `select` and a dyn-cast kind of thing, such that the diagnostic doesn't print the name if one doesn't exist:

`S.Diag(...) << /*The select:*/isa<NamedDecl>(OffendingDecl) << /*the name: */dyn_cast<NamedDecl>(OffendingDecl)`

It has to be a dyn-cast even though we know it won't be null, because we don't have a way of short-circuiting via 'select'.

Reply:
That is a good idea, will implement this in the next push.